lib/Target/TfheRustBool/TfheRustBoolEmitter.cpp

```diff
@@ -165,29 +168,70 @@ void TfheRustBoolEmitter::emitAssignPrefix(Value result) {
   os << "let " << variableNames->getNameForValue(result) << " = ";
 }
 
+void TfheRustBoolEmitter::emitReferenceConversion(Value value) {
+  auto varName = variableNames->getNameForValue(value);
+  os << "let " << varName << "_ref = " << varName << ".clone();\n";
+  os << "let " << varName << "_ref: Vec<&Ciphertext> = " << varName
+     << ".iter().collect();\n";
+}
+
 LogicalResult TfheRustBoolEmitter::printSksMethod(
     ::mlir::Value result, ::mlir::Value sks, ::mlir::ValueRange nonSksOperands,
     std::string_view op, SmallVector<std::string> operandTypes) {
-  emitAssignPrefix(result);
-
-  auto operandTypesIt = operandTypes.begin();
-  os << variableNames->getNameForValue(sks) << "." << op << "(";
-  os << commaSeparatedValues(nonSksOperands, [&](Value value) {
-    auto *prefix = value.getType().hasTrait<PassByReference>() ? "&" : "";
-    // First check if a DefiningOp exists
-    // if not: comes from function definition
-    mlir::Operation *op = value.getDefiningOp();
-    if (op) {
-      prefix = isa<tensor::ExtractOp>(op) ? "" : prefix;
-    } else {
-      prefix = "";
+  if (isa<TensorType>(nonSksOperands[0].getType())) {
+    auto *opParent = nonSksOperands[0].getDefiningOp();
+    if (!opParent) {
+      for (auto nonSksOperand : nonSksOperands) {
+        emitReferenceConversion(nonSksOperand);
+      }
     }
 
-    return prefix + variableNames->getNameForValue(value) +
-           (!operandTypes.empty() ? " as " + *operandTypesIt++ : "");
-  });
-  os << ");\n";
-  return success();
+    emitAssignPrefix(result);
+
+    os << variableNames->getNameForValue(sks) << "." << op << "_packed(";
+    os << commaSeparatedValues(
+        {nonSksOperands[0], nonSksOperands[1]}, [&](Value value) {
+          auto *prefix = value.getType().hasTrait<PassByReference>() ? "&" : "";
+          auto suffix = "";
+          // First check if a DefiningOp exists
+          // if not: comes from function definition
+          mlir::Operation *opParent = value.getDefiningOp();
+          if (opParent) {
+            prefix = isa<tensor::ExtractOp>(opParent) ? prefix : "";
+            prefix =
+                isa<tensor::FromElementsOp>(value.getDefiningOp()) ? "&" : "";
+          } else {
+            prefix = "&";
+            suffix = "_ref";
+          }
+
+          return prefix + variableNames->getNameForValue(value) + suffix;
+        });
+    os << ");\n";
+    return success();
+
+  } else {
+    emitAssignPrefix(result);
+
+    auto operandTypesIt = operandTypes.begin();
+    os << variableNames->getNameForValue(sks) << "." << op << "(";
+    os << commaSeparatedValues(nonSksOperands, [&](Value value) {
+      auto *prefix = value.getType().hasTrait<PassByReference>() ? "&" : "";
+      // First check if a DefiningOp exists
+      // if not: comes from function definition
```

Comment:
Optional: If you want to be extra careful about this: you would cast it to a BlockArgument and check that the owner is a func op.